workspaces/src/worker/impls.rs

```diff
@@ -0,0 +1,163 @@
+use near_primitives::borsh::BorshSerialize;
+use near_primitives::types::{AccountId, Balance, FunctionArgs, StoreKey};
+
+use crate::network::{
+    Account, AllowDevAccountCreation, CallExecution, CallExecutionDetails, Contract, NetworkClient,
+    NetworkInfo, StatePatcher, TopLevelAccountCreator,
+};
+use crate::rpc::client::Client;
+use crate::worker::Worker;
+
+unsafe impl<T> Send for Worker<T> where T: Send {}
+unsafe impl<T> Sync for Worker<T> where T: Sync {}
+
+impl<T> Clone for Worker<T> {
+    fn clone(&self) -> Self {
+        Self {
+            workspace: self.workspace.clone(),
+        }
+    }
+}
+
+impl<T> AllowDevAccountCreation for Worker<T> where T: AllowDevAccountCreation {}
+
+#[async_trait]
+impl<T> TopLevelAccountCreator for Worker<T>
+where
+    T: TopLevelAccountCreator + Send + Sync,
+{
+    async fn create_tla(
+        &self,
+        id: AccountId,
+        signer: InMemorySigner,
+    ) -> anyhow::Result<CallExecution<Account>> {
+        self.workspace.create_tla(id, signer).await
+    }
+
+    async fn create_tla_and_deploy<P: AsRef<Path> + Send + Sync>(
+        &self,
+        id: AccountId,
+        signer: InMemorySigner,
+        wasm: P,
+    ) -> anyhow::Result<CallExecution<Contract>> {
+        self.workspace.create_tla_and_deploy(id, signer, wasm).await
+    }
+}
+
+impl<T> NetworkInfo for Worker<T>
+where
+    T: NetworkInfo,
+{
+    fn name(&self) -> String {
+        self.workspace.name()
+    }
+
+    fn root_account_id(&self) -> AccountId {
+        self.workspace.root_account_id()
+    }
+
+    fn keystore_path(&self) -> std::path::PathBuf {
+        self.workspace.keystore_path()
+    }
+
+    fn rpc_url(&self) -> String {
+        self.workspace.rpc_url()
+    }
+
+    fn helper_url(&self) -> String {
+        self.workspace.helper_url()
+    }
+}
+
+#[async_trait]
+impl<T> StatePatcher for Worker<T>
+where
+    T: StatePatcher + Send + Sync,
+{
+    async fn patch_state<U>(
+        &self,
+        contract_id: AccountId,
+        key: String,
+        value: &U,
+    ) -> anyhow::Result<()>
+    where
+        U: BorshSerialize + Send + Sync,
+    {
+        self.workspace.patch_state(contract_id, key, value).await
+    }
+}
+
+impl<T> Worker<T>
+where
+    T: NetworkClient,
+{
+    pub(crate) fn client(&self) -> &Client {
+        self.workspace.client()
+    }
+
+    pub async fn call(
+        &self,
+        contract: &Contract,
+        method: String,
+        args: Vec<u8>,
+        deposit: Option<Balance>,
+    ) -> anyhow::Result<CallExecutionDetails> {
+        self.client()
+            .call(
+                contract.signer(),
+                contract.id(),
+                method,
+                args,
+                None,
+                deposit,
+            )
+            .await
+            .map(Into::into)
+    }
+
+    pub async fn view(
+        &self,
+        contract_id: AccountId,
+        method_name: String,
+        args: FunctionArgs,
```

Comment:
I feel strongly that we shouldn't be exposing types from `nearcore` in public API. Crates like `near-primitives` do not currently have public APIs and are woefully inadequate for use in public, stable APIs with backwards compatibility guarantees. 

Copy-pasting data definitions from nearcore is quite some amount of busy work. But without doing that, you'll be forced to break *clients* of workspaces-rs every time impl details of nearcore change.


In this specific case, additionally, it doesnt' make sense that call uses Vec<u8> while view uses `FunctionArgs`.

Reply:
Ah, this was also one of my TODOs, but that fell through as I was rushing it for thanksgiving. This one adds quite a bit of types, so let me make a separate issue out of this and we can fix it up after this PR. Sounds good to you?

Reply:
addressed in https://github.com/near/workspaces-rs/pull/25